addons/l10n_ro_efactura/models/account_move.py

```diff
@@ -0,0 +1,130 @@
+        self.ensure_one()
+        return f"{self.name.replace('/', '_')}.xml"
+
+    def _l10n_ro_edi_compute_errors(self, xml_data):
+        """ Compute possible errors before sending E-Factura """
+        self.ensure_one()
+        errors = []
+        if not self.company_id.l10n_ro_edi_access_token:
+            errors.append(_('Romanian access token not found. Please generate or fill it in the settings.'))
+        if not xml_data:
+            errors.append(_('CIUS-RO XML attachment not found.'))
+        return errors
+
+    def _l10n_ro_edi_send_invoice(self, xml_data):
+        """ Called by _call_web_service_after_invoice_pdf_render in account.move.send.
+            The state flow of Romanian E-Factura invoice goes as follows:
+             - Pre-check any errors from invoice's company before sending, return if found any
+             - Send to E-Factura, and based on the result:
+                 - if error -> delete all other error documents and create a new error document
+                 - if success -> delete all other sending documents and create a new sending document"""
+        self.ensure_one()
+        if errors := self.company_id._l10n_ro_edi_get_errors_pre_request():
+            self.env['l10n_ro_edi.document']._create_document_invoice_sending_failed(self, '\n'.join(errors))
+            return
+
+        result = self.env['l10n_ro_edi.document']._request_ciusro_send_invoice(
+            company=self.company_id,
+            xml_data=xml_data,
+            move_type=self.move_type,
+        )
+        if 'error' in result:
+            self.l10n_ro_edi_document_ids.filtered(lambda d: d.state == 'invoice_sending_failed').unlink()
+            self.env['l10n_ro_edi.document']._create_document_invoice_sending_failed(self, result['error'])
+        else:
+            self.l10n_ro_edi_document_ids.filtered(lambda d: d.state == 'invoice_sending').unlink()
+            self.env['l10n_ro_edi.document']._create_document_invoice_sending(self, result['key_loading'])
+
+    def _l10n_ro_edi_fetch_invoice_sending_documents(self):
+        """ Collects all selected active documents in self and process them as a batch.
+            Make a fetch request for each document. Based on the received result,
+            if error -> generate error document on that document's invoice
+            else -> immediately make a download request and process it
+        """
+        documents = self.l10n_ro_edi_active_document_id.filtered(lambda d: d.state == 'invoice_sending')
+        if not documents:
+            return
+        session = requests.session()
+        to_delete_documents = self.env['l10n_ro_edi.document']
+
+        for document in documents:
+            invoice = document.invoice_id
+            if errors := invoice.company_id._l10n_ro_edi_get_errors_pre_request():
+                to_delete_documents |= invoice.l10n_ro_edi_document_ids.filtered(lambda d: d.state == 'invoice_sending_failed')
```

Comment:
Your method seems to have a n² complexity. What about:
```py
for invoice in self.filtered(lambda inv: inv.l10n_ro_state == 'invoice_sending'):
    ...
```
... because anything you need seems stored on the invoice.